File: scripts/wikidata.py
#! /usr/bin/env nix-shell
#! nix-shell -i python -p "python38.withPackages(ps: with ps; [ requests ])""

import json
import requests
import sys
from subprocess import run, PIPE

query = sys.argv[1]
baseURL = "https://www.wikidata.org/w/api.php"
params = {"action": "wbsearchentities",
          "search": query,
          "language": "en",
          "format": "json"} 
response = requests.get(baseURL, params=params)

if response.ok:
    decoded = json.loads(response.text)
    # print(decoded['search'][0])
    itemsWithDescriptions = '\n'.join(['\t'.join(["http:"+item.get('url'), item.get('label', ''), item.get('descriptions', '')])
                             for item in decoded['search']])
    out = run(['rofi', '-dmenu'], input=itemsWithDescriptions, encoding='ascii', capture_output=True)
    print('#+wikidata: '+out.stdout.split('\t')[0])
else:
    print("Something went wrong.")
    print(response)

File: scripts/wikidata.hs
#! /usr/bin/env nix-shell
#! nix-shell -i runhaskell -p "haskellPackages.ghcWithPackages (ps: with ps; [http-conduit http-client-tls yaml])"

{-# LANGUAGE OverloadedStrings #-}
import           Data.Aeson            (Value, Array, Object, (.:))
import qualified Data.ByteString.Char8 as L8
import           Data.ByteString.Internal as BS
import           Network.HTTP.Simple
import           System.Environment
-- import qualified Data.Yaml             as Yaml



main :: IO ()
main = do
    args <- getArgs
    let query = Just $ BS.packChars $ head args
    let request
            = setRequestQueryString [("action", Just "wbsearchentities")
                                    ,("search", query)
                                    ,("language", Just "en")
                                    ,("format", Just "json")]
            "https://www.wikidata.org/w/api.php"
    response <- httpJSON request
    let body = (getResponseBody response) :: Object
    let result = body .: "search"
    print result
    -- L8.putStrLn $ Yaml.encode result
